src/streams.h

```diff
@@ -520,6 +520,10 @@ class BufferedFile
             throw std::ios_base::failure("Rewind limit must be less than buffer size");
     }
 
+    ~BufferedFile() { fclose(); }
```

Comment:
It is better to check the return value of `fclose()`, otherwise an IO error may remain unnoticed and unreported, leading to silent data corruption.

The problem with calling `fclose()` from the destructor is that there is no way to signal a failure to the caller - the destructor does not return a value, does not take arguments and is not supposed to `throw` either.

`fclose()` from a destructor is a design issue, better avoid that. https://github.com/bitcoin/bitcoin/pull/29307 is related.

Reply:
BufferedFile is only used for reading, so I don't think fclose can fail in a meaningful way? #29307 also ignores the result...

Reply:
> BufferedFile is only used for reading

Is it possible to enforce that in some way? Also that the parent `AutoFile` has been used only for reading prior to passing it to `BufferedFile`?

> https://github.com/bitcoin/bitcoin/pull/29307 also ignores the result

#29307 fixes/improves some things, but not that

Reply:
I see that now you explicitly close the file. Maybe assert there like:

```cpp
assert(opt_buffered_file->fclose() == 0);
```
and remove this destructor?

Reply:
Is it not the case that now this destructor is not needed because you explicitly call `fclose()`?